### DISCUSSION TITLE
Call method with extensionObject as Argument

### DISCUSSION THREAD

**Discussion by BenediktEnger** [FreeOpcUa/opcua-asyncio#946] [Q&A] [answered]
Hi,

I'm trying to call a method which has a ExtensionObject as Input Argument. Unfortunally, I can't find out, how to pass this argument to the method *call_method()* the right way. 
The problem is, I don't know how to get the variantType of the object. 
I checked the extension objects stored after calling *load_data_type_definitions()* but the object I'm looking for is not part of it.

Strange is: I have Siemens PLC OPC UA Server and the structure I'm looking for is nested in a generic structure:
![image](https://user-images.githubusercontent.com/44339476/175024690-c5c4b6e0-3ae6-42c8-b467-1c73aedeca49.png)

I hope somebody can help me. Thanks!

**Comment by schroeder-** [comment, marked as answer]
This is how it should work:
```Python
arg1 = ua.SK2_AmsMeasurementStatusType()
obj.call_method(method_id, arg1)
```

If load_data_type_definitions does not return the node for those class we holes in load_data_type_definitions.
Which version are you using? Because I did some fixes for load_data_type_definitions in master (#901)
Also is there something in the logging?

**Reply by BenediktEnger** [reply]
I tried the latest release but it is still the same. 
The only structure I can see under the extensionObjects is "SimaticVersion" which you can see on the image.
```
ua.SK2_StartDriveMeasurement_V1InputArgument()
```
returns an error, that this attribute is not available.

I attached the log file. I can see, that the client browsed the structure "SK2_StartDriveMeasurement_V1InputArgument".
Maybe the client has a problem, because there are " in the nodeID.

[log.txt](https://github.com/FreeOpcUa/opcua-asyncio/files/8964937/log.txt)
![NodeID](https://user-images.githubusercontent.com/44339476/175240068-11d73a11-0338-41ca-b1d4-8a57b4ed7521.PNG)

**Reply by schroeder-** [reply]
#901 is not released yet, you could try current master.

**Reply by BenediktEnger** [reply]
Problem still extists. When I load a custom nodeset file, the types can be detected. 
Unfortunally, in my use case I can't use a nodeset file.

**Reply by schroeder-** [reply]
Can you try #949?
In some cases a Subtype of a DataType isn't browsed and added.

**Reply by BenediktEnger** [reply]
This seems to work. When I call `client.load_data_type_definitions()`, the structures are read.
The problem is now, that in one structure I use a String. Don't ask me why, but Siemens is not using the String node from namespace 0, but they are creating there own Node:
![image](https://user-images.githubusercontent.com/44339476/175889030-c78a60e1-4c1e-4d7c-a73e-704f1e2314af.png)

Because it is no extension object, it will not be found and the following code crashes. (structures104.py line 186)
```
if sfield.DataType.NamespaceIndex == 0 and sfield.DataType.Identifier in ua.ObjectIdNames:
            if sfield.DataType.Identifier == 24:
                uatype = "Variant"
            elif sfield.DataType.Identifier == 22:
                uatype = "ExtensionObject"
            else:
                uatype = ua.ObjectIdNames[sfield.DataType.Identifier]
        elif sfield.DataType in ua.extension_objects_by_datatype:
            uatype = ua.extension_objects_by_datatype[sfield.DataType].__name__
        elif sfield.DataType in ua.enums_by_datatype:
            uatype = ua.enums_by_datatype[sfield.DataType].__name__
        else:
            # FIXME: we are probably missing many custom tyes here based on builtin types
            # maybe we can use ua_utils.get_base_data_type()
            raise RuntimeError(f"Unknown datatype for field: {sfield} in structure:{struct_name}, please report")
```

When I read the FIXME note, the problem seems to be known. Unfortunally, the described function seems not to exist.
Can you help me with this problem? Thanks

**Reply by schroeder-** [reply]
In the current master this code is alread reworked:
https://github.com/FreeOpcUa/opcua-asyncio/blob/0d006fc6978138f0fe9b56fc1f2489fcfa4d5d05/asyncua/common/structures104.py#L197-L206
As #949 is merged into master both fixes are in master now.

**Reply by BenediktEnger** [reply]
Thanks. The current master works. 
But there where two issues I had to fix: 

1. The function `register_basetype` is writing the class_type as value at `basetype_by_datatype`.
 This is not working because when getting the default value of the attribute (`default_value = get_default_value(uatype)` at strucutres104.py l. 211), a name is expected.

```def register_basetype(name, nodeid, class_type):
    """
    Register a new allias of basetypes for automatic decoding and make them available in ua module
    """
    logger.info("registring new basetype alias: %s %s %s", name, nodeid, class_type)
    basetype_by_datatype[nodeid] = name #NOT class_type
    basetype_datatypes[class_type] = nodeid
    import asyncua.ua

    setattr(asyncua.ua, name, class_type)
```

2. When creating the code at strucutures104.py l. 179, the node id has to be sorrounded by ' instead of " because there are " sometimes in the node  id:
```
 code = f"""

@dataclass
class {struct_name}{base_class}:

    '''
    {struct_name} structure autogenerated from StructureDefinition object
    '''

    data_type = ua.NodeId.from_string('{data_type.to_string()}')

"""
```

**Reply by schroeder-** [reply]
Problem 1 is a quick fix no problem.

The issue with quotes in opc ua names is tricky see issue #783, there is a outstandig pr that tried to fix it: #784

**Reply by schroeder-** [reply]
Created #954 for register_basetype
